test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/Helpers/TestTimeoutStream.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace NuGet.Protocol.FuncTest.Helpers
+{
+    internal class TestTimeoutStream : Stream
+    {
+        private Stream _stream;
+
+        public TestTimeoutStream(Stream stream)
+        {
+            _stream = stream;
+        }
+
+        public override bool CanRead => _stream.CanRead;
+
+        public override bool CanSeek => _stream.CanSeek;
+
+        public override bool CanWrite => _stream.CanWrite;
+
+        public override long Length => _stream.Length;
+
+        public override long Position
+        {
+            get => _stream.Position;
+            set => _stream.Position = value;
+        }
+
+        public override void Flush()
+        {
+            _stream.Flush();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            // Don't time out immediately, because we want to test what happens when some data is sent successfully,
+            // then the timeout happens, and NuGet.Protocol has to retry and handle the previously partial data.
+            var read = _stream.Read(buffer, offset, count);
+            if (read > 0)
+            {
+                return read;
+            }
+
+            throw new TimeoutException();
```

Comment:
It looks not 1st try. Exactly when and how often hit this?

Reply:
Either you answered your own question in https://github.com/NuGet/NuGet.Client/pull/4425#discussion_r797844091, or I don't understand your question.